src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -1201,31 +1201,50 @@ public static SqlConnectionStringBuilder CreateConnectionStringBuilder(Connectio
             ChangeDatabaseParams changeDatabaseParams,
             RequestContext<bool> requestContext)
         {
-            await requestContext.SendResult(ChangeConnectionDatabaseContext(changeDatabaseParams.OwnerUri, changeDatabaseParams.NewDatabase));
+            await requestContext.SendResult(ChangeConnectionDatabaseContext(changeDatabaseParams.OwnerUri, changeDatabaseParams.NewDatabase, true));
         }
 
         /// <summary>
         /// Change the database context of a connection.
         /// </summary>
         /// <param name="ownerUri">URI of the owner of the connection</param>
         /// <param name="newDatabaseName">Name of the database to change the connection to</param>
-        public bool ChangeConnectionDatabaseContext(string ownerUri, string newDatabaseName)
+        public bool ChangeConnectionDatabaseContext(string ownerUri, string newDatabaseName, bool force = false)
         {
             ConnectionInfo info;
             if (TryFindConnection(ownerUri, out info))
             {
                 try
                 {
-                    foreach (DbConnection connection in info.AllConnections)
+                    info.ConnectionDetails.DatabaseName = newDatabaseName;
+
+                    foreach (string key in info.AllConnectionTypes)
                     {
-                        if (connection.State == ConnectionState.Open)
+                        DbConnection conn;
+                        info.TryGetConnection(key, out conn);
+                        if (conn != null && conn.State == ConnectionState.Open)
                         {
-                            connection.ChangeDatabase(newDatabaseName);
+                            if (info.IsCloud && force)
+                            {
+                                conn.Close();

```

Comment:
Can you check whether the connection is already targeting the new database? The reason is that in the scenario where Query Execution changes DB, one connection will already have been changed (via `USE Database`) to target the new DB but the default and any other connections won't. 

Ideally, we only close & reopen connections that don't already target the new database. This will make the code more robust and able to handle other callers who want to make use of it.

Reply:
fixed